### PR TITLE
Ad hoc implementation of progress bar

### PR DESCRIPTION
**Pull request by mrgreenhub**
The requested feature of a progress bar for reducing the output of info messages in the cli.

### CODE REVIEW

File: bin/hr/hr.go [hunk 1]
```diff
@@ -351,6 +366,7 @@ func main() {
 	pflag.BoolVar(&conv.formatter.TinyFormat, "tiny", false, "use penlog hr-tiny format")
 	pflag.StringVarP(&prioLevelRaw, "priority", "p", "debug", "show messages with a lower priority level")
 	pflag.StringArrayVarP(&filterSpecs, "filter", "f", []string{}, "write logs to a file with filters")
+	pflag.BoolVar(&conv.showProgress, "show-progress", false, "replace the consecutive displayment of info messages with a progress bar")
```
Comment (rumpelsepp) [inline bin/hr/hr.go:369]:
```suggestion
	pflag.BoolVar(&conv.showProgress, "progress", false, "replace the consecutive displayment of info messages with a progress bar")
```

Comment (rumpelsepp) [reply bin/hr/hr.go:369]: Hmm. Not sure if this naming is correct. It is not a progress bar. It is more a `--collapse` or something.

File: bin/hr/hr.go [hunk 2]
```diff
@@ -235,7 +241,16 @@ func (c *converter) transform(r io.Reader) {
 			}
 		}
 		if hrLine, err := c.formatter.Format(d); err == nil {
-			fmt.Println(hrLine)
+			if isatty(uintptr(syscall.Stdout)) {
+				// If in progress mode override infos in the same line
+				if priority == penlog.PrioInfo && c.showProgress {
+					fmt.Print("\033[2K" + hrLine + "\r")
```
Comment (rumpelsepp) [inline bin/hr/hr.go:247]: Not a fan of hardcoded escape codes. Let's rename the `color.go` file to `escapes.go` and add this one there. Maybe a function `eraseLine()` makes sense here.

Comment (mrgreenhub) [reply bin/hr/hr.go:247]: In principle II'm with you on this one.
I just didn't want to change any filenames or make a new one just for this single code.

File: bin/hr/hr.go [hunk 3]
```diff
@@ -220,9 +221,14 @@ func (c *converter) transform(r io.Reader) {
 				continue
 			}
 		}
+
+		var (
```
Comment (rumpelsepp) [inline bin/hr/hr.go:225]: Please collapse in one line `var foo type`

File: bin/hr/hr.go (hunk 1 above)
Comment (mrgreenhub) [reply bin/hr/hr.go:369]: I stcked to @peckto's naming but I agree

File: bin/hr/hr.go (hunk 2 above)
Comment (rumpelsepp) [reply bin/hr/hr.go:247]: I figure there might come more escapes in the future. So let's do it.

Comment (rumpelsepp) [reply bin/hr/hr.go:247]: Ajo, let's prefer `fmt.Printf()` to `fmt.Print()` and concatenating strings.

File: bin/hr/hr.go (hunk 3 above)
Comment (mrgreenhub) [reply bin/hr/hr.go:225]: ok

File: bin/hr/hr.go (hunk 1 above)
Comment (rumpelsepp) [reply bin/hr/hr.go:369]: I know, I don't like my `--collapse` suggestion very well. But it fits a bit more to the actual thing.

Comment (mrgreenhub) [reply bin/hr/hr.go]: In the new version, it's called volatile-infos, maybe that's a better fititng name.

Comment (rumpelsepp) [reply bin/hr/hr.go]: last complain: what about `--volatile` (without the `-info`)?

**Comment by rumpelsepp** [comment]
I'm fine. Could you add a test for this? Maybe just [here](https://github.com/Fraunhofer-AISEC/penlog/blob/master/tests/cli/0001-hr-check.bats), you might reuse some of the logging artifacts as well.

**Comment by mrgreenhub** [comment]
By the way:
This only overwrites the last line in the terminal.
If an info output spans multiple lines, then a prefix is still visible in the lines above.
One thing which I tried out is to save `\0337` and restore `\0338` the cursor, and to clear everything underneath the cursor `\033[J`, which seemed to be a promising way to handle multiline output.
However, this is only better than a bare carriage return in case the terminal has not scrolled in between.
Further research also yielded, that there is probably no way to achieve this using just bare terminal escape sequences.
So overall, I sticked to the simple and more consistent solution using carriage return and a line clear.

**Comment by rumpelsepp** [comment]
> If an info output spans multiple lines, then a prefix is still visible in the lines above.

Good point. But I figure that info entries spanning mulitple lines are just too much. We could just disable the feature in case of multiline entries and just print the damn thing.

**Comment by mrgreenhub** [comment]
> > If an info output spans multiple lines, then a prefix is still visible in the lines above.
> 
> Good point. But I figure that info entries spanning mulitple lines are just too much. We could just disable the feature in case of multiline entries and just print the damn thing.

Well, maybe there is a little misunderstanding here.
A mutliline output in my previous comment means, that the output spans mutliple line on the terminal, no matter if it actually contains a newline or not.
So in the code, we actually do not know when an output spans multiple lines and therefore cannot disable the feature.
If we would know, when the output will result in multiple lines, then we might even be able to clear them properly in the first place.

**Comment by rumpelsepp** [comment]
> A mutliline output in my previous comment means, that the output spans mutliple lines in the terminal, no matter if it actually contains a newline or not.

In this case in my opinion it is ok to accept broken output, since the user could just resize the terminal.

**Comment by mrgreenhub** [comment]
> I'm fine. Could you add a test for this? Maybe just [here](https://github.com/Fraunhofer-AISEC/penlog/blob/master/tests/cli/0001-hr-check.bats), you might reuse some of the logging artifacts as well.

In principle yes.
In practice not really.
Because this functionality is guarded by a check of isatty(), this is probably never taking effect in the test cases, right?.

**Comment by rumpelsepp** [comment]
> Because this functionality is guarded by a check of isatty(), this is probably never taking effect in the test cases, right?.

Fuck. For the colors, I introduced `PENLOG_FORCE_COLORS` or so. This case here unfortunate, since you might never want to enforce this behaviour. But IIRC you could allocate a pty and it should work as expected. We might steal [this](https://unix.stackexchange.com/a/231355). :D

**Comment by mrgreenhub** [comment]
But IIRC you could allocate a pty and it should work as expected. We might steal [this](https://unix.stackexchange.com/a/231355). :D

Seems a little overkill to me

**Comment by mrgreenhub** [comment]
> > Because this functionality is guarded by a check of isatty(), this is probably never taking effect in the test cases, right?.
> 
> Fuck. For the colors, I introduced `PENLOG_FORCE_COLORS` or so. This case here unfortunate, since you might never want to enforce this behaviour. But IIRC you could allocate a pty and it should work as expected. We might steal [this](https://unix.stackexchange.com/a/231355). :D

Seems a little overkill

**Comment by rumpelsepp** [comment]
maybe or maybe not. let's skip the test for now. is this ready?

**Comment by mrgreenhub** [comment]
From my point of view it now is.

**Comment by rumpelsepp** [comment]
Currently testing this. Is cool.

**Comment by rumpelsepp** [comment]
merged

**Pull request closed without merging**